### PR TITLE
Ftruncate fixes

### PR DESCRIPTION
**Pull request by keithbostic**
Two fixes when using ftruncate as the file-extension call:

* The ftruncate call can't set WT_FH.size, otherwise we'll skip those blocks when allocating blocks from the file,
* The file size has to be read inside of the lock being used to avoid races with writers.

Reference #1871.

### CODE REVIEW

**Comment by keithbostic** [comment]
@sueloverso, would you please pick up the review on this one?

My principle worries are:
* the locking relationship between `__wt_block_extend` and `__wt_block_write_off`,
* the calculation as to whether or not a particular thread should extend the file at the start of `__wt_block_extend`,
* and, if the change in f26f1c1 (where the underlying `ftruncate` call no longer sets `fh->size` and `fh->extend_size`), won't affect the logging calls that potentially call `ftruncate`. (I looked at them, and I don't see problems, but I'm not familiar enough with that code to be confident I'm correct.)

Thank you!

File: src/block/block_write.c [hunk 1]
```diff
@@ -21,6 +21,124 @@ __wt_block_header(WT_BLOCK *block)
...
+__wt_block_extend(WT_SESSION_IMPL *session, WT_BLOCK *block,
+    WT_FH *fh, wt_off_t offset, size_t align_size, int *release_lockp)
+{
+	WT_DECL_RET;
+	int locked;
+
+	/*
+	 * The locking in this function is messy: the live system is locked when
+	 * we're called, by definition, but that lock may have been acquired by
+	 * our our caller or our caller's caller. If it's our caller's lock and
```
Comment (sueloverso) [inline src/block/block_write.c]: "our our"

File: src/block/block_write.c [hunk 2]
```diff
@@ -21,6 +21,124 @@ __wt_block_header(WT_BLOCK *block)
...
+	 * we can unlock it before returning (either before extending the file
+	 * or afterward, depending on the call used), then release_lock is set.
```
Comment (sueloverso) [inline src/block/block_write.c]: then release_lock is cleared.

File: src/block/block_write.c [hunk 3]
```diff
@@ -21,6 +21,124 @@ __wt_block_header(WT_BLOCK *block)
...
+	 *
+	 * If we unlock, but then find out we need a lock after all, re-acquire
+	 * the lock (and set release_lock so our caller knows to release it).
+	 */
+	locked = 1;
```
Comment (sueloverso) [inline src/block/block_write.c]: Looking at the calling code, I know *release_lockp is set when called, but it seems like we should really do *release_lockp = 1 here just to pair it properly with the local knowledge that we're locked.  (Or maybe an assert.)

**Comment by sueloverso** [comment]
@keithbostic There were a few minor comments, but otherwise this looks good.  I studied the logging code and it syncs/closes the fh after the truncate calls.  I stared at the calculation a while and it looks correct.

File: src/block/block_write.c (hunk 1 above)
Comment (keithbostic) [inline src/block/block_write.c]: Fixed.

File: src/block/block_write.c (hunk 2 above)
Comment (keithbostic) [inline src/block/block_write.c]: I think this is correct.

File: src/block/block_write.c (hunk 3 above)
Comment (keithbostic) [inline src/block/block_write.c]: I think this is correct.

**Pull request merged**